snapfaas/src/fs/mod.rs

```diff
@@ -344,1687 +56,713 @@ impl From<reqwest::Method> for HttpVerb {
+    /// accumulated.
+    pub fn to_invokable<B: BackingStore>(&self, fs: &FS<B>) -> DirectGate {
+        let mut cur = self.get(fs).unwrap().unlabel().clone();
+        let mut privilege = Component::dc_true();
+        let mut declassify = Component::dc_true();
+        let mut invoker_integrity_clearance = Component::dc_true();
+        loop {
+            match cur {
+                Gate::Direct(gate) => {
+                    privilege = privilege & gate.privilege;
+                    invoker_integrity_clearance = invoker_integrity_clearance & gate.invoker_integrity_clearance;
+                    return DirectGate {
+                        privilege,
+                        invoker_integrity_clearance,
+                        declassify,
+                        function: gate.function,
                     }
-                }
-                DirEntry::File(file) => {
-                    let _ = visited.insert(file.object_id);
-                }
-                DirEntry::Gate(gate) => {
-                    let _ = visited.insert(gate.object_id);
-                }
-                DirEntry::Blob(blob) => {
-                    let _ = visited.insert(blob.object_id);
-                }
-                DirEntry::Service(service) => {
-                    let _ = visited.insert(service.object_id);
+                },
+                Gate::Redirect(redirect_gate) => {
+                    privilege = privilege & redirect_gate.privilege;
+                    invoker_integrity_clearance = invoker_integrity_clearance & redirect_gate.invoker_integrity_clearance;
+                    declassify = declassify & redirect_gate.declassify;
+                    cur = redirect_gate.gate.get(fs).unwrap().unlabel().clone();
                 }
             }
         }
+    }
 
-        let diff = objects.difference(&visited).map(|&x| x).collect::<Vec<_>>();
-        for obj in diff.iter() {
-            self.storage.del(&obj.to_be_bytes());
+    pub fn replace<B: BackingStore>(&self, new_gate: Gate, fs: &FS<B>) -> Result<(), FsError> {
+        {
+            let new_priv = match &new_gate {
+                Gate::Direct(d) => &d.privilege,
+                Gate::Redirect(r) => &r.privilege,
+            };
+            PRIVILEGE.with(|privilege| {
+                let privilege = privilege.borrow();
+                if !privilege.implies(new_priv) {
+                    Err(FsError::PrivilegeError(PrivilegeError::CannotDelegate))
+                } else {
+                    Ok(())
+                }
+            })?;
         }
-
-        Ok(diff)
+        let mut gate = self.get(fs).unwrap();
+        gate.write(new_gate)?;
+        Ok(self.set(&gate, &fs.0))
     }
 }
 
-impl Directory {
-    pub fn label(&self) -> &Buckle {
-        &self.label
-    }
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct DirectGate {
+    pub privilege: Component,
+    pub invoker_integrity_clearance: Component,
+    pub declassify: Component,
```

Comment:
Ah, I think it's never used, but if memory serves my thought was to allow a gate to declassify the post-computation by the provided component. This would be equivalent to the function code explicitly declassifying, except that both redirects and direct gates can include these.

To clarify, this is not actually used anywhere at the moment, so doesn't have this affect. We can remove or we can actually use that. Worth (before merging) figuring out if this is good/useful to include.